google/cloud/storage/client.h

```diff
@@ -97,7 +99,7 @@ class Client {
   BucketMetadata GetBucketMetadata(std::string const& bucket_name,
                                    Modifiers&&... modifiers) {
     internal::GetBucketMetadataRequest request(bucket_name);
-    request.set_multiple_parameters(std::forward<Modifiers>(modifiers)...);
+    request.set_multiple_modifiers(std::forward<Modifiers>(modifiers)...);
```

Comment:
It'd be helpful to clearly call out the difference between header fields versus query parameters.

At a first glance I find that it helps to read `set_multiple_request_parameters` or `set_multiple_request_options` (second is borrowed from the Java client library).

One open question is how to determine source and destination values during use-case of rewrite operation with a CSK? As the operation requires a source encryption key and destination encryption key.  Looking at [existing samples](https://cloud.google.com/storage/docs/encryption/using-customer-supplied-keys#rotating) for more insight on existing libraries.

Reply:
> It'd be helpful to clearly call out the difference between header fields versus query parameters.

Helpful how and to whom?  For the user all this stuff is immaterial, and I think being able to pass optional stuff uniformly makes their life easier.  For us, the library developers, yes, it would be nicer to keep them separate in more obvious ways than just their base class (query parameters inherit from `WellKnownParameter`, query headers inherit from `WellKnownHeader`).

I tend to favor ease of use over use of implementation in these cases.

> At a first glance I find that it helps to read `set_multiple_request_parameters` or `set_multiple_request_options` (second is borrowed from the Java client library).

Note that in this design the application developer calls neither.  The application developer calls:

```C++
client.GetBucketMetadata(
    bucket, Param1(...), Param2(...), ParamThatHappensToBeHeader3(...));
```

We call `set_multiple_parameters()` (now `set_multiple_modifiers()`) under the hood, but the user should not care.  The classes in question (e.g. `GetBucketMetadataRequest`) are in the `internal::` namespace, that signals that we do not expect (and we discourage) users from using those classes directly.

> One open question is how to determine source and destination values during use-case of rewrite operation with a CSK? As the operation requires a source encryption key and destination encryption key. Looking at existing samples for more insight on existing libraries.

That would be:

```C++
client.Copy(
    destination_bucket, destination_object,
    source_bucket, source_object,
    SourceEncryptionKey(algo, key, sha),
    DestinationEncryptionKey(algo2, key2, sha2));
```

The fact that those keys are encoded as headers is kind of an implementation detail.

Reply:
@frankyn I just re-read this:

> Helpful how and to whom?

And it came out too harsh, I really just did not follow what audience you had in mind, my phrasing was kind of nasty, my sincere apologies.

Reply:
Np! I misunderstood during my first pass. Given the PR changes being in the underlying implementation this sounds good to me.

Reply:
What are your thoughts on using `option` instead of `modifier`?